File: demo/KSYLiveDemo/KSYPlayerDemo/KSYUIRecorderKit.m
//
//  KSYUIRecorderKit.m
//  playerRecorder
//
//  Created by ksyun on 16/10/26.
//  Copyright © 2016年 ksyun. All rights reserved.
//

#import "KSYUIRecorderKit.h"

#import <GPUImage/GPUImage.h>
#import <libksygpulive/KSYGPUPicMixer.h>
#import <libksygpulive/KSYAudioMixer.h>
#import <libksygpulive/KSYGPUPicOutput.h>
#import <libksygpulive/KSYMoviePlayerController.h>
#import <libksygpulive/KSYDummyAudioSource.h>



@interface KSYUIRecorderKit()
{
    BOOL _bBackground;
}
//图层0 player
@property  GPUImageTextureInput* textureInput;
@property  NSInteger playerLayer;
//图层1 UI
@property  GPUImageUIElement* uiElementInput;
@property  NSInteger uiLayer;
//伟大的混流器
@property  KSYGPUPicMixer* uiMixer;
@property  KSYAudioMixer *aMixer;

@property  KSYDummyAudioSource* dummyAudio;
@property (nonatomic, readonly) int dummyTrack;
@property (nonatomic, readonly) int playerTrack;

@property  KSYGPUPicOutput* gpuToStr;
@property  CMTime lastPts;

@end

@implementation KSYUIRecorderKit

- (instancetype)init {
    if (!(self = [super init]))
    {
        return nil;
    }
    
    _bPlayRecord = NO;
    CGSize sz = [UIScreen mainScreen].bounds.size;
    _contentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, sz.width,sz.height)];
    _contentView.backgroundColor = [UIColor clearColor];
    _contentView.contentScaleFactor = 2;
    
    _playerLayer = 0 ;
    _uiLayer = 1;
    
    _dummyTrack = 0;
    _playerTrack = 1;
    
    _bBackground = NO;
    
    [self createMixer];
    
    [self createWriter];
    
    [self registerApplicationObservers];

    return self;
}

-(void)dealloc{
    if(_contentView)
        _contentView = nil;
    
    if(_uiElementInput)
        _uiElementInput = nil;
    
    if(_uiMixer){
        [_uiMixer clearPicOfLayer:_playerLayer];
        [_uiMixer clearPicOfLayer:_uiLayer];
        _uiMixer = nil;
    }
    
    if(_aMixer)
        _aMixer = nil;
    
    if(_gpuToStr)
        _gpuToStr = nil;
    
    if(_writer){
        [_writer stopWrite];
        _writer = nil;
    }
    
    if(_dummyAudio){
        [_dummyAudio stop];
        _dummyAudio = nil;
    }
    
    [self unregisterApplicationObservers];
}

-(void)startRecord:(NSURL*) path
{
    [_writer startWrite:path];
    _bPlayRecord = YES;
}

-(void)stopRecord
{
    [_writer stopWrite];
    _bPlayRecord = NO;
}

-(void)createWriter{
    _gpuToStr = [[KSYGPUPicOutput alloc] initWithOutFmt:kCVPixelFormatType_32BGRA];
    _writer = [[KSYMovieWriter alloc] init];
    _writer.videoCodec = KSYVideoCodec_VT264;
    _writer.audioCodec = KSYAudioCodec_AAC;
    _writer.bWithVideo = YES;
    _writer.bWithAudio = YES;
    
    __weak KSYUIRecorderKit * weakKit = self;
    _aMixer.audioProcessingCallback = ^(CMSampleBufferRef buf){        
        [weakKit.writer processAudioSampleBuffer:buf];
    };
    
    _gpuToStr.videoProcessingCallback = ^(CVPixelBufferRef pixelBuffer, CMTime timeInfo){
        [weakKit.writer processVideoPixelBuffer:pixelBuffer timeInfo:timeInfo];
    };
    
    [_uiMixer addTarget:_gpuToStr];
}

-(void)createMixer{
    _uiMixer = [[KSYGPUPicMixer alloc] init];
    _uiMixer.masterLayer = _uiLayer;
    [_uiMixer setPicRect:CGRectMake(-1,-1,0.0,0.0) ofLayer:_playerLayer];
    [_uiMixer setPicRotation:kGPUImageFlipVertical ofLayer:_playerLayer];
    [_uiMixer setPicAlpha:1.0 ofLayer:_playerLayer];
    
    [_uiMixer setPicRect:CGRectMake(0,0,1.0,1.0) ofLayer:_uiLayer];
    [_uiMixer setPicAlpha:1.0 ofLayer:_uiLayer];
    
    _aMixer = [[KSYAudioMixer alloc]init];
    _aMixer.mainTrack = _dummyTrack;
    
    [_aMixer setTrack:_dummyTrack enable:YES];
    [_aMixer setTrack:_playerTrack enable:YES];
    
    AudioStreamBasicDescription format;
    memset(&format, 0, sizeof(format));
    format.mSampleRate = 44100;
    format.mFormatID = kAudioFormatLinearPCM;
    format.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked;
    format.mChannelsPerFrame = 2;
    format.mBitsPerChannel = 16;
    format.mBytesPerFrame = (format.mBitsPerChannel * format.mChannelsPerFrame / 8);
    format.mFramesPerPacket = 1;
    format.mBytesPerPacket = format.mBytesPerFrame * format.mFramesPerPacket;
    
    __weak KSYUIRecorderKit *weakKit = self;
    _dummyAudio = [[KSYDummyAudioSource alloc]initWithAudioFmt:format];
    _dummyAudio.audioProcessingCallback = ^(CMSampleBufferRef sampleBuffer){
        [weakKit.aMixer processAudioSampleBuffer:sampleBuffer of:weakKit.dummyTrack];
        weakKit.lastPts = CMSampleBufferGetPresentationTimeStamp(sampleBuffer);        
    };
    CMTime startPts = CMTimeMake((int64_t)(0 * 1000), 1000);
    [_dummyAudio startAt:startPts];
}

-(void) processAudioSampleBuffer:(CMSampleBufferRef) buf
{
    @synchronized (self) {
        if(!_bPlayRecord)
            return;
        
        [_aMixer processAudioSampleBuffer:buf of:_playerTrack];
    }
}

-(void) processWithTextureId:(GLuint)InputTexture
                 TextureSize:(CGSize)TextureSize
                        Time:(CMTime)time
{
    @synchronized (self) {
        if(!_bPlayRecord && _bBackground)
            return;
        
        if(_uiMixer)
        {
            _textureInput = [[GPUImageTextureInput alloc]initWithTexture:InputTexture size:TextureSize];
            [_uiMixer clearPicOfLayer:_playerLayer];
            [_textureInput addTarget:_uiMixer atTextureLocation:_playerLayer];
            [_textureInput processTextureWithFrameTime:time];
            
            _uiElementInput = [[GPUImageUIElement alloc] initWithView:_contentView];
            [_uiMixer clearPicOfLayer:_uiLayer];
            [_uiElementInput addTarget:_uiMixer atTextureLocation:_uiLayer];
            [_uiElementInput updateWithTimestamp:_lastPts];
        }
    }
}

- (void)registerApplicationObservers
{
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationWillEnterForeground)
                                                 name:UIApplicationWillEnterForegroundNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationDidBecomeActive)
                                                 name:UIApplicationDidBecomeActiveNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationWillResignActive)
                                                 name:UIApplicationWillResignActiveNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationDidEnterBackground)
                                                 name:UIApplicationDidEnterBackgroundNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(applicationWillTerminate)
                                                 name:UIApplicationWillTerminateNotification
                                               object:nil];
}

- (void)unregisterApplicationObservers
{    
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                        name:UIApplicationWillEnterForegroundNotification
                                                      object:nil];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIApplicationDidBecomeActiveNotification
                                                  object:nil];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIApplicationWillResignActiveNotification
                                                  object:nil];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIApplicationDidEnterBackgroundNotification
                                                  object:nil];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIApplicationWillTerminateNotification
                                                  object:nil];
}

- (void)applicationWillEnterForeground
{
    _bBackground = NO;
}

- (void)applicationDidBecomeActive
{
    _bBackground = NO;
}

- (void)applicationWillResignActive
{
    _bBackground = YES;
}

- (void)applicationDidEnterBackground
{
    _bBackground = YES;
}

- (void)applicationWillTerminate
{
    _bBackground = YES;
}

@end

File: demo/KSYLiveDemo_Swift/KSYPlayerDemo_Swift/KSYUIRecorderKit.swift
//
//  KSYUIRecorderKit.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 2017/1/23.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

class KSYUIRecorderKit: NSObject {
    /* ui图层*/
    var contentView: UIView!
    /*是否开启录屏*/
    var bPlayRecord: Bool = false
    /*录制视频文件*/
    var writer: KSYMovieWriter?
    
    private
    // player
    var textureInput: GPUImageTextureInput?
    var playerLayer: Int = 0
    // UI
    var uiElementInput: GPUImageUIElement?
    var uiLayer: Int = 0
    // mixer
    var uiMixer: KSYGPUPicMixer?
    var aMixer: KSYAudioMixer?
    
    var dummyAudio: KSYDummyAudioSource?
    var dummyTrack: Int = 0
    var playerTrack: Int = 0
    
    var gpuToStr: KSYGPUPicOutput?
    var lastPts: CMTime = CMTime()
    
    var bBackground: Bool = false
    var processQueue: DispatchQueue = DispatchQueue.init(label: "com.ksyun.recordProcessQueue")
    
    override init() {
        super.init()
        bPlayRecord = false
        contentView = UIView.init(frame: CGRect.init(x: 0,
                                                     y: 0,
                                                     width: UIScreen.main.bounds.width,
                                                     height: UIScreen.main.bounds.height))
        contentView.backgroundColor = .clear
        
        playerLayer = 0
        uiLayer = 1
        
        dummyTrack = 0
        playerTrack = 1
        
        bBackground = false
        
        createMixer()
        
        createWriter()
        
        registerApplicationObservers()
    }
    
    deinit {
        contentView = nil
        uiElementInput = nil
        if let _ = uiMixer {
            uiMixer!.clearPic(ofLayer: playerLayer)
            uiMixer!.clearPic(ofLayer: uiLayer)
            uiMixer = nil
        }
        
        aMixer = nil
        gpuToStr = nil
        
        if let _ = writer {
            writer!.stopWrite()
            writer = nil
        }
        
        if let _ = dummyAudio {
            dummyAudio!.stop()
            dummyAudio = nil
        }
        unregisterApplicationObservers()
    }
    
    func startRecord(path: URL) {
        processQueue.sync {
            writer?.startWrite(path)
            bPlayRecord = true
        }
    }
    
    func stopRecord() {
        processQueue.sync {
            writer?.stopWrite()
            bPlayRecord = false
        }
    }
    
    func createWriter() {
        gpuToStr = KSYGPUPicOutput.init(outFmt: kCVPixelFormatType_32BGRA)
        writer = KSYMovieWriter()
        writer?.videoCodec = KSYVideoCodec.VT264
        writer?.audioCodec = KSYAudioCodec.AAC
        writer?.bWithVideo = true
        writer?.bWithAudio = true
        
        aMixer?.audioProcessingCallback = { [weak self] (buf) -> Void in
            self?.writer?.processAudioSampleBuffer(buf)
        }
        
        gpuToStr?.videoProcessingCallback = { [weak self] (pixelBuffer, timeInfo) -> Void in
            self?.writer?.processVideoPixelBuffer(pixelBuffer, timeInfo: timeInfo)
        }
        
        uiMixer?.addTarget(gpuToStr!)
    }
    
    func createMixer() {
        uiMixer = KSYGPUPicMixer()
        uiMixer!.masterLayer = UInt(uiLayer)
        uiMixer!.setPicRect(CGRect.init(x: -1, y: -1, width: 0, height: 0), ofLayer: playerLayer)
        uiMixer!.setPicRotation(kGPUImageFlipVertical, ofLayer: playerLayer)
        uiMixer!.setPicAlpha(1.0, ofLayer: playerLayer)
        
        uiMixer!.setPicRect(CGRect.init(x: 0, y: 0, width: 1, height: 1), ofLayer: uiLayer)
        uiMixer!.setPicAlpha(1.0, ofLayer: uiLayer)
        
        aMixer = KSYAudioMixer()
        aMixer!.mainTrack = Int32(dummyTrack)
        
        aMixer!.setTrack(Int32(dummyTrack), enable: true)
        aMixer!.setTrack(Int32(playerTrack), enable: true)
        
        var format: AudioStreamBasicDescription = AudioStreamBasicDescription()
        memset(&format, 0, MemoryLayout.size(ofValue: format))
        format.mSampleRate = 44100
        format.mFormatID = kAudioFormatLinearPCM
        format.mFormatFlags = kLinearPCMFormatFlagIsSignedInteger | kLinearPCMFormatFlagIsPacked
        format.mChannelsPerFrame = 2
        format.mBitsPerChannel = 16
        format.mBytesPerFrame = format.mBitsPerChannel * format.mChannelsPerFrame / 8
        format.mFramesPerPacket = 1
        format.mBytesPerPacket = format.mBytesPerFrame * format.mFramesPerPacket
        
        dummyAudio = KSYDummyAudioSource.init(audioFmt: format)
        dummyAudio!.audioProcessingCallback = { [weak self] (sampleBuffer) -> Void in
            self?.aMixer?.processAudioSampleBuffer(sampleBuffer, of: Int32(self?.dummyTrack ?? 0))
            self?.lastPts = CMSampleBufferGetPresentationTimeStamp(sampleBuffer!)
        }
        let startPts = CMTime.init(value: 0, timescale: 1000)
        dummyAudio?.start(at: startPts)
    }

    func processAudioSampleBuffer(buf: CMSampleBuffer) {
        processQueue.sync {
            if !bPlayRecord {
                return
            }
            
            aMixer?.processAudioSampleBuffer(buf, of: Int32(playerTrack))
        }
    }
    
    func process(with inputTexture: GLuint, textureSize: CGSize, time: CMTime) {
        processQueue.sync {
            if !bPlayRecord && bBackground {
                return
            }
            
            if let _ = uiMixer {
                textureInput = GPUImageTextureInput.init(texture: inputTexture, size: textureSize)
                uiMixer!.clearPic(ofLayer: playerLayer)
                textureInput!.addTarget(uiMixer!, atTextureLocation: playerLayer)
                textureInput!.processTexture(withFrameTime: time)
                
                uiElementInput = GPUImageUIElement.init(view: contentView)
                uiMixer!.clearPic(ofLayer: uiLayer)
                uiElementInput!.addTarget(uiMixer!, atTextureLocation: uiLayer)
                uiElementInput!.update(withTimestamp: lastPts)
            }
        }
    }

    func registerApplicationObservers() {
        NotificationCenter.default.addObserver(self, selector: #selector(applicationWillEnterForeground), name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationDidBecomeActive), name: NSNotification.Name.UIApplicationDidBecomeActive, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationWillResignActive), name: NSNotification.Name.UIApplicationWillResignActive, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationDidEnterBackground), name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationWillTerminate), name: NSNotification.Name.UIApplicationWillTerminate, object: nil)
    }
    
    func unregisterApplicationObservers() {
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationDidBecomeActive, object: nil)
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationWillResignActive, object: nil)
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationWillTerminate, object: nil)
    }
    
    func applicationWillEnterForeground() {
        bBackground = false
    }
    
    func applicationDidBecomeActive() {
        bBackground = false
    }
    
    func applicationWillResignActive() {
        bBackground = true
    }

    func applicationDidEnterBackground() {
        bBackground = true
    }
    
    func applicationWillTerminate() {
        bBackground = true
    }
}
